util/cache/cache.go

```diff
@@ -1,21 +1,198 @@
 package cache
 
 import (
-	"errors"
+	"encoding/json"
+	"fmt"
 	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/spf13/cobra"
+
+	"github.com/argoproj/argo-cd"
+	"github.com/argoproj/argo-cd/common"
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util/hash"
+)
+
+const (
+	defaultCacheExpiration          = 24 * time.Hour
+	connectionStatusCacheExpiration = 1 * time.Hour
+	appStateCacheExpiration         = 1 * time.Hour
+	repoCacheExpiration             = 24 * time.Hour
+	oidcCacheExpiration             = 3 * time.Minute
 )
 
-var ErrCacheMiss = errors.New("cache: key is missing")
+type OIDCState struct {
+	// ReturnURL is the URL in which to redirect a user back to after completing an OAuth2 login
+	ReturnURL string `json:"returnURL"`
+}
+
+// NewCache creates new instance of Cache
+func NewCache(cacheClient CacheClient) *Cache {
+	return &Cache{client: cacheClient}
+}
+
+// AddCacheFlagsToCmd adds flags which control caching to the specified command
+func AddCacheFlagsToCmd(cmd *cobra.Command) func() (*Cache, error) {
+	redisAddress := ""
+	sentinelAddresses := make([]string, 0)
+	sentinelMaster := ""
+
+	cmd.Flags().StringVar(&redisAddress, "redis", "", "Redis server hostname and port (e.g. argocd-redis:6379). ")
+	cmd.Flags().StringArrayVar(&sentinelAddresses, "sentinel", []string{}, "Redis sentinel hostname and port (e.g. argocd-redis-ha-announce-0:6379). ")
+	cmd.Flags().StringVar(&sentinelMaster, "sentinelmaster", "master", "Redis sentinel master group name.")
+	return func() (*Cache, error) {
+		if len(sentinelAddresses) > 0 {
+			client := redis.NewFailoverClient(&redis.FailoverOptions{
+				MasterName:    sentinelMaster,
+				SentinelAddrs: sentinelAddresses,
+			})
+			return NewCache(NewRedisCache(client, defaultCacheExpiration)), nil
+		}
+
+		if redisAddress == "" {
+			redisAddress = common.DefaultRedisAddr
+		}
+		client := redis.NewClient(&redis.Options{
+			Addr:     redisAddress,
+			Password: "",
+			DB:       0,
+		})
+		return NewCache(NewRedisCache(client, defaultCacheExpiration)), nil
+	}
+}
+
+// Cache provides strongly types methods to store and retrieve values from shared cache
+type Cache struct {
+	client CacheClient
+}
+
+func appManagedResourcesKey(appName string) string {
+	return fmt.Sprintf("app|managed-resources|%s", appName)
```

Comment:
I think we should consider prefixing something like a db "version" to the keys so that when we make changes to the responses, we bump the version and the cache is automatically invalidated

Reply:
This is already implemented in `getItem`/`setItem` methods. As we agreed offline switched to separete const for cache version.